docs/src/main/sphinx/develop/tests.md

```diff
@@ -0,0 +1,424 @@
+# Testing guidelines
+
+The Trino project aims to provide high quality artifacts and features to all
+users. A suite of tests used for development, continuous integration builds,
+and releases is critical for achieving this goal.
+
+The following requirements and guidelines help contributors to create tests with
+the following goals:
+
+* Fast execution.
+* Low requirements in terms of hardware, software, and compute power, and
+  therefore low cost.
+* Reliable execution and test results across multiple invocations.
+* Complex enough to test necessary aspects, while avoiding unnecessary
+  complexity.
+* Enable quick, iterative development on developer workstation.
+
+## Requirements
+
+A number of requirements apply to all new tests, and any refactoring work of
+existing tests:
+
+* All tests must use JUnit 5.
+* All tests must use statically imported AssertJ assertions, typically from
+  `org.assertj.core.api.Assertions`.
+* Test class names must start with `Test`, for example `TestExample`
+* Test classes should be defined as package-private and final.
+* Test method must start with `test`, for example `testExplain()`
+* Test methods should be defined as package-private.
+* Tests must be written as unit tests when possible, product or other
+  integration tests should be avoided.
+* Tests must not be duplicated across unit and product tests, or different
+  plugins and other integrations.
+
+## Guidelines
+
+In addition to the requirements, a number of specific guidelines help with
+authoring new tests as well as refactoring existing tests.
+
+### Focus on high value tests
+
+Testing in Trino is extremely expensive, and slows down all development as they
```

Comment:
I feel the "why" needs to be expanded a bit. This is mostly due to the nature of the project, where Trino supports multiple very different data sources where each connectors is extensively tested. We only run relevant tests in PRs based on impacted Maven modules, but all tests are always executed after pushing any changes to the master branch.